File: solutions/typescript/0074.ts
/**
 * 74. 搜索二维矩阵
 * 把二维矩阵视作一维有序数组，使用二分查找
 */
function searchMatrix(matrix: number[][], target: number): boolean {
    if (matrix.length === 0 || matrix[0].length === 0) {
        return false;
    }

    const m = matrix.length;
    const n = matrix[0].length;
    let left = 0;
    let right = m * n - 1;

    while (left <= right) {
        const mid = (left + right) >>> 1;
        const row = Math.floor(mid / n);
        const col = mid % n;
        const val = matrix[row][col];

        if (val === target) {
            return true;
        } else if (val < target) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }

    return false;
}

export default searchMatrix;

File: solutions/rust/0074.rs
impl Solution {
    /// 74. 搜索二维矩阵
    /// 把二维矩阵视作一维有序数组，使用二分查找
    pub fn search_matrix(matrix: Vec<Vec<i32>>, target: i32) -> bool {
        if matrix.is_empty() || matrix[0].is_empty() {
            return false;
        }

        let n = matrix[0].len();
        let total = matrix.len() * n;
        let mut left = 0usize;
        let mut right = total;

        while left < right {
            let mid = left + (right - left) / 2;
            let (row, col) = (mid / n, mid % n);
            let val = matrix[row][col];

            if val < target {
                left = mid + 1;
            } else {
                right = mid;
            }
        }

        left < total && matrix[left / n][left % n] == target
    }
}
